<a href="https://colab.research.google.com/github/KeatonDahya/LING229/blob/main/Assignment_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Short Program and Written Reflection 2
Keaton Dahya

300571027

In [1]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [4]:
import os

# Define folder paths for Google Drive
horror_path = '/content/drive/My Drive/HorrorBooks'
romance_path = '/content/drive/My Drive/RomanceBooks'

# Function to load text files from a folder
def load_texts_from_folder(folder_path):
    texts = []
    for filename in os.listdir(folder_path):
        if filename.endswith('.txt'):  # Ensure only .txt files are loaded
            with open(os.path.join(folder_path, filename), 'r', encoding='utf-8') as file:
                texts.append(file.read())
    return texts

horror_texts = load_texts_from_folder(horror_path)
romance_texts = load_texts_from_folder(romance_path)

# **Introduction**

Research Question- Do horror novels employ a higher frequency of fear-related vocabulary compared to romance novels, which may favor more positively-valenced emotional words?

The purpose of this corpus is to compare two distinct categories of texts: Horror novels and Romance novels. The comparison is aimed at analyzing linguistic features such as vocabulary diversity, word frequency, and tone to identify differences in the writing style and emotional focus of these genres.

# What Are the Texts About?

Horror Novels:
The horror texts in this corpus explore themes of fear, suspense, and the supernatural. These novels often use darker language, vivid imagery, and negative emotional vocabulary to create tension and evoke fear in readers.

Romance Novels:
The romance texts focus on themes of love, relationships, and emotional connection. They typically use positive, sentimental, and descriptive language to highlight intimacy and human connection.

# Where do they come From?
The texts have been sourced from Project Gutenberg, a digital library of free eBooks in the public domain. Each novel was downloaded as a .txt file and saved into two separate folders:

HorrorBooks - contains 10 horror novels.

RomanceBooks - contains 10 romance novels.

# Categories and How They Were Defined

The two categories, Horror and Romance, were defined based on the primary themes and genres of the texts:

Texts categorized as Horror include elements of fear, supernatural events, and suspense.

Texts categorized as Romance focus on relationships, emotional intimacy, and positive resolution.

# How Were the Texts Collected?

Step 1: Searched for well-known horror and romance novels on Project Gutenberg.

Step 2: Downloaded each novel in .txt format and saved them into their
respective folders.

Step 3: Ensured there were 10 texts per category to provide a balanced comparison.


In [2]:
import os
import nltk
from nltk.tokenize import word_tokenize
import pandas as pd
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

# **Total Tokens and Total Types**

Length: Romance novels contain almost double the total tokens compared to horror novels. This reflects the structural nature of each genre:

Romance: The genre emphasizes long, intricate emotional journeys, detailed descriptions of relationships, and often includes dialogue-heavy sections, which increase word count. Additionally, romance novels frequently explore personal growth and multiple perspectives, requiring longer narratives.

Horror: The focus in horror is often on brevity and intensity to sustain suspense and fear. Horror novels are more concise and employ sharper pacing to keep readers engaged, which limits overall length.

Vocabulary Size: Romance novels also contain 50% more unique words (types),
reflecting their broader thematic scope and diverse emotional vocabulary.

Romance: Words related to emotions (e.g., love, longing, heartbreak) and detailed descriptions contribute to this larger vocabulary.

Horror: The vocabulary focuses on vivid imagery, unsettling descriptions, and unique adjectives to create an atmosphere of fear. However, the scope is narrower as horror stories typically revolve around specific events or settings.


---


# **Tokens and Types Per Text**

Romance novels show greater variability in text length, with the largest text exceeding 250,000 tokens. This highlights the genre's flexibility in narrative scope – romance novels often span long periods, include multiple character arcs, and contain extensive dialogue.

Horror novels are more consistent in length, with most texts being shorter. This consistency reflects the genre's focus on pacing – shorter, tighter narratives sustain tension and fear more effectively.

In romance novels, the higher range of types per text suggests a wider vocabulary usage for detailed and emotionally rich prose.

In horror novels, even with fewer tokens, the variety of types is still significant, emphasizing the use of descriptive and chilling language.

This supports the idea that horror writers rely on evocative and precise word choices to build suspense within shorter texts.


---


# **Average Tokens and Types per Text**

Romance novels, on average, have almost double the word count per text compared to horror. This reflects:

The emotional depth and complexity of romance novels, which often explore relationships over extended timelines.

Extensive dialogue, backstories, and intricate settings, which naturally add more words.

The higher average unique words per text in romance novels shows the need for diverse vocabulary to capture nuances in emotions, relationships, and settings. Romance novels are filled with emotionally descriptive words.

Horror's Focus on Precision: Despite having fewer total words, horror novels still maintain a significant vocabulary size. This suggests that horror authors are more deliberate in word choice to maximize the emotional impact in fewer words.


---


# **Lexical Diversity**
**Horror Novels: 0.0315**

**Romance Novels: 0.0257**

Lexical Diversity measures the ratio of unique words to total words. A higher value means more variety in vocabulary relative to the total word count.
Horror novels show slightly higher lexical diversity than romance novels (0.0315 vs. 0.0257).

Horror Novels: Higher lexical diversity indicates a need for rich and varied descriptions to create suspense, fear, and vivid imagery. Horror authors aim for precision to paint unsettling mental images in fewer words.

Romance Novels: Lower lexical diversity is not a negative factor. Romance novels often rely on repetitive language for emotional emphasis and rhythm.

In [5]:
# Function to calculate metrics
def calculate_metrics(texts):
    total_tokens = 0
    unique_tokens = set()
    tokens_per_text = []
    types_per_text = []

    for text in texts:
        tokens = word_tokenize(text.lower())
        tokens_per_text.append(len(tokens))  # Tokens in this text
        types_per_text.append(len(set(tokens)))  # Unique types in this text
        total_tokens += len(tokens)
        unique_tokens.update(tokens)

    avg_tokens = sum(tokens_per_text) / len(tokens_per_text)
    avg_types = sum(types_per_text) / len(types_per_text)
    lexical_diversity = len(unique_tokens) / total_tokens

    return {
        "Total tokens (entire corpus)": total_tokens,
        "Total types (entire corpus)": len(unique_tokens),
        "Tokens per text": tokens_per_text,
        "Types per text": types_per_text,
        "Average tokens per text": round(avg_tokens, 2),
        "Average types per text": round(avg_types, 2),
        "Lexical diversity": round(lexical_diversity, 4)
    }

# Load texts and calculate metrics for Horror
horror_texts = load_texts_from_folder(horror_path)
horror_metrics = calculate_metrics(horror_texts)

# Print results for Horror
print("\n**Horror Category Metrics:**")
for key, value in horror_metrics.items():
    print(f"{key}: {value}")


**Horror Category Metrics:**
Total tokens (entire corpus): 762484
Total types (entire corpus): 24007
Tokens per text: [103874, 129548, 71466, 102701, 66480, 91590, 40191, 59940, 67098, 29596]
Types per text: [8095, 9081, 7365, 8796, 5060, 8104, 4623, 5785, 6722, 4167]
Average tokens per text: 76248.4
Average types per text: 6779.8
Lexical diversity: 0.0315


In [6]:
# Function to load text files from a folder
def load_texts_from_folder(folder_path):
    texts = []
    for filename in os.listdir(folder_path):
        if filename.endswith('.txt'):
            with open(os.path.join(folder_path, filename), 'r', encoding='utf-8') as file:
                texts.append(file.read())
    return texts

# Function to calculate metrics
def calculate_metrics(texts):
    total_tokens = 0
    unique_tokens = set()
    tokens_per_text = []
    types_per_text = []

    for text in texts:
        tokens = word_tokenize(text.lower())
        tokens_per_text.append(len(tokens))  # Tokens in this text
        types_per_text.append(len(set(tokens)))  # Unique types in this text
        total_tokens += len(tokens)
        unique_tokens.update(tokens)

    avg_tokens = sum(tokens_per_text) / len(tokens_per_text)
    avg_types = sum(types_per_text) / len(types_per_text)
    lexical_diversity = len(unique_tokens) / total_tokens

    return {
        "Total tokens (entire corpus)": total_tokens,
        "Total types (entire corpus)": len(unique_tokens),
        "Tokens per text": tokens_per_text,
        "Types per text": types_per_text,
        "Average tokens per text": round(avg_tokens, 2),
        "Average types per text": round(avg_types, 2),
        "Lexical diversity": round(lexical_diversity, 4)
    }

# Load texts and calculate metrics for Romance
romance_texts = load_texts_from_folder(romance_path)
romance_metrics = calculate_metrics(romance_texts)

print("\n**Romance Category Metrics:**")
for key, value in romance_metrics.items():
    print(f"{key}: {value}")


**Romance Category Metrics:**
Total tokens (entire corpus): 1393360
Total types (entire corpus): 35804
Tokens per text: [65701, 197465, 251840, 89205, 105285, 147833, 225984, 104855, 78867, 126325]
Types per text: [5535, 12507, 12810, 7753, 9979, 10661, 13250, 8100, 6924, 7257]
Average tokens per text: 139336.0
Average types per text: 9477.6
Lexical diversity: 0.0257


# **Lexical Profiles**

To construct the lexical profiles for my texts (Horror Novels vs. Romance Novels), I will analyze each text’s word frequency, word variety, and overall richness. This analysis provides insights into how language is used differently across genres. I will calculate:

**Tokens**

The total number of words in a text, including repetitions. Tokens reflect the overall length and verbosity of the text.

Texts with higher token counts are typically longer and more descriptive, while shorter texts may focus on conciseness and impact.

I will use the nltk.word_tokenize() function to split each text into individual words (tokens). Lowercasing all words ensures consistency and avoids duplicates caused by capitalization.


---


**Types**

The number of unique words in the text. Types indicate the vocabulary diversity and word variation within the text.

A text with a high number of unique words (types) often explores complex ideas or employs a varied vocabulary, as seen in genres like romance novels.

I will extract unique words using the set() function, then calculate their total count.


---


**Lexical Diversity**

The ratio of unique words (types) to the total number of words (tokens)

Lexical Diversity = Unique Types / Total Tokens

 Lexical diversity provides a measure of language richness. Horror novels, for instance, might use concise yet impactful language, leading to higher diversity in fewer words. Romance novels, with longer narratives and repetitive emotional vocabulary, may show lower diversity.

I will divide the unique types by the total tokens to compute this ratio.




In [ ]:
# Function to calculate lexical profile for a single text
def calculate_lexical_profile(text):
    tokens = word_tokenize(text.lower())  # Tokenize and lowercase text
    total_tokens = len(tokens)            # Total words (tokens)
    unique_types = len(set(tokens))       # Unique words (types)
    lexical_diversity = unique_types / total_tokens  # Lexical diversity

    return total_tokens, unique_types, lexical_diversity

# Function to process a folder of texts and calculate profiles
def process_folder(folder_path):
    results = []
    for file_name in os.listdir(folder_path):
        if file_name.endswith('.txt'):  # Only process .txt files
            with open(os.path.join(folder_path, file_name), 'r', encoding='utf-8') as file:
                text = file.read()
                total_tokens, unique_types, lexical_diversity = calculate_lexical_profile(text)
                results.append({
                    'File': file_name,
                    'Total Tokens': total_tokens,
                    'Unique Types': unique_types,
                    'Lexical Diversity': round(lexical_diversity, 4)
                })
    return results

horror_path = '/content/drive/MyDrive/HorrorBooks'
romance_path = '/content/drive/MyDrive/RomanceBooks'

horror_results = process_folder(horror_path)
romance_results = process_folder(romance_path)

horror_df = pd.DataFrame(horror_results)
romance_df = pd.DataFrame(romance_results)

horror_averages = {
    'File': 'AVERAGE',
    'Total Tokens': horror_df['Total Tokens'].mean(),
    'Unique Types': horror_df['Unique Types'].mean(),
    'Lexical Diversity': horror_df['Lexical Diversity'].mean()
}

romance_averages = {
    'File': 'AVERAGE',
    'Total Tokens': romance_df['Total Tokens'].mean(),
    'Unique Types': romance_df['Unique Types'].mean(),
    'Lexical Diversity': romance_df['Lexical Diversity'].mean()
}

horror_df = pd.concat([horror_df, pd.DataFrame([horror_averages])], ignore_index=True)
romance_df = pd.concat([romance_df, pd.DataFrame([romance_averages])], ignore_index=True)

print("=== Horror Category Metrics ===")
print(horror_df)
print("\n=== Romance Category Metrics ===")
print(romance_df)

# Save results to CSV files
horror_df.to_csv('/content/drive/MyDrive/horror_lexical_profile.csv', index=False)
romance_df.to_csv('/content/drive/MyDrive/romance_lexical_profile.csv', index=False)



=== Horror Category Metrics ===
            File  Total Tokens  Unique Types  Lexical Diversity
0    horror1.txt      103874.0        8095.0            0.07790
1    horror2.txt      129548.0        9081.0            0.07010
2    horror3.txt       71466.0        7365.0            0.10310
3    horror4.txt      102701.0        8796.0            0.08560
4    horror5.txt       66480.0        5060.0            0.07610
5    horror6.txt       91590.0        8104.0            0.08850
6    horror7.txt       40191.0        4623.0            0.11500
7    horror8.txt       59940.0        5785.0            0.09650
8    horror9.txt       67098.0        6722.0            0.10020
9   horror10.txt       29596.0        4167.0            0.14080
10       AVERAGE       76248.4        6779.8            0.09538

=== Romance Category Metrics ===
             File  Total Tokens  Unique Types  Lexical Diversity
0    romance1.txt       65701.0        5535.0            0.08420
1    romance2.txt      197465.0     

# Interpretation of Results

My research question asked whether horror novels employ a higher frequency of fear-related vocabulary compared to romance novels, which may favor more positively-valenced emotional words. While our initial analysis didn't directly measure emotional vocabulary frequencies, our lexical analysis reveals several interesting patterns that provide insight into how these genres differ linguistically.

Structural Differences
The most striking difference is in text length: romance novels are significantly longer, averaging 139,336 words compared to horror's 76,248 words. This substantial difference (romance novels being almost twice as long) suggests different storytelling approaches. Romance novels tend to develop emotional relationships gradually through extended character development and dialogue, while horror novels favor brevity and intensity to maintain suspense.

Vocabulary Usage and Diversity
Though horror novels are shorter, they demonstrate higher lexical diversity (0.0315) compared to romance novels (0.0257). This higher diversity suggests horror authors employ a more varied vocabulary despite using fewer words overall. This finding partially supports our hypothesis about genre-specific vocabulary usage - horror writers appear to carefully select diverse words to create precise emotional effects, possibly including fear-related terms, rather than relying on repetitive language.

Implications for Genre Conventions
Romance novels' lower lexical diversity doesn't indicate less sophisticated writing, but rather reflects genre conventions. The repetition of emotional terms may serve to reinforce themes of love and relationship development. Conversely, horror's higher lexical diversity suggests authors may need a broader vocabulary to maintain tension and create varied descriptions of fear-inducing scenarios.

Future Research Directions
To directly address our original research question, future analysis should:

Create specific lexicons of fear-related and love-related terms to measure their frequencies

Implement sentiment analysis to compare emotional valence between genres

Examine specific passages where emotional vocabulary clusters occur

Analyze adjective usage specifically, as these might carry significant
emotional weight

Compare subgenres (e.g., supernatural horror vs. psychological horror, historical romance vs. contemporary romance) to identify more nuanced patterns

These findings suggest that while both genres deal with intense emotions, they employ distinctly different linguistic strategies to achieve their effects. Horror novels use precise, varied vocabulary in shorter texts, while romance novels develop emotional resonance through longer narratives with more repetitive but emotionally consistent language.


